WrightTools/artists/_interact.py

```diff
@@ -0,0 +1,381 @@
+        sp_x.set_ylim(-1.1, 1.1)
+        sp_y.set_xlim(-1.1, 1.1)
+
+    def update_sideplot_slices():
+        xlim = ax0.get_xlim()
+        ylim = ax0.get_ylim()
+        x0 = current_state.xpos
+        y0 = current_state.ypos
+
+        crosshair_hline.set_data(np.array([xlim, [y0, y0]]))
+        crosshair_vline.set_data(np.array([[x0, x0], ylim]))
+
+        arr = channel[current_state.slices].squeeze()
+        if wt_kit.get_index(data.axes, xaxis) < wt_kit.get_index(data.axes, yaxis):
+            arr = arr.T.copy()
+
+        x_temp = np.abs(xaxis.points - x0)
+        x_index = np.argmin(x_temp)
+        side_plot = arr[:, x_index].copy()
+        side_plot = norm(side_plot, channel.signed)
+        line_sp_y.set_data(side_plot, yaxis.points)
+
+        y_temp = np.abs(yaxis.points - y0)
+        y_index = np.argmin(y_temp)
+        side_plot = arr[y_index].copy()
+        side_plot = norm(side_plot, channel.signed)
+        line_sp_x.set_data(xaxis.points, side_plot)
+
+    def update_local(index):
+        slices = get_slices(sliders, data.axes, verbose=verbose)
+        if verbose:
+            print(index)
+        if radio.value_selected.strip() == "global":
+            current_state.local = False
+        if radio.value_selected.strip() == "local":
+            current_state.local = True
+        arr = channel[slices].squeeze()
+        clim = get_clim(channel, current_state.local, arr)
+        print(clim)
+        ticklabels = gen_ticklabels(np.linspace(*clim, 11))
+        colorbar.set_ticklabels(ticklabels)
+        obj2D.set_clim(*clim)
+        fig.canvas.draw_idle()
+
+    def update(info):
+        # is info a value?  then we have a slider
+        # is info an object with xydata?  then we have an event
+        slices = get_slices(sliders, data.axes, verbose=verbose)
+        if slices != current_state.slices:  # a Slider moved; need to update all plot objects
+            arr = channel[slices].squeeze()
+            if wt_kit.get_index(data.axes, xaxis) < wt_kit.get_index(data.axes, yaxis):
+                arr = arr.T.copy()
+            # TODO: why am I stripping off array information?
+            # cf. https://stackoverflow.com/questions/29009743
+            obj2D.set_array(arr[:-1, :-1].ravel())
+            clim = get_clim(channel, current_state.local, arr)
+            obj2D.set_clim(*clim)
+            ticklabels = gen_ticklabels(np.linspace(*clim, 11))
+            colorbar.set_ticklabels(ticklabels)
+            sp_x.collections.clear()
+            sp_y.collections.clear()
+            draw_sideplot_projections(arr)
+            current_state.slices = slices
+            if line_sp_x.get_visible() and line_sp_y.get_visible():
+                update_sideplot_slices()
+        elif info.inaxes == ax0:  # crosshairs
+            x0 = info.xdata
+            y0 = info.ydata
+            if x0 is None or y0 is None:
+                print(info, info.xydata)
+                raise AttributeError
```

Comment:
please use a WrightTools derived exception

Reply:
I would say either `TypeError`

`AttributeError` tends to specifically mean the call to `__getattr__` (which is called by doing `foo.bar`) fails. That is not the case here, as the attribute of info is defined (or at least does not raise an exception)
(Python docs)[https://docs.python.org/3.7/library/exceptions.html#AttributeError]

TypeError is appropriate as x0/y0 have NoneType, not the type expected (I beleive numeric, maybe float specifically). We define in exceptions.py a TypeError which inherits from both our own base exception and the standard library TypeError, so it can be caught by a calling function by either.

ValueError is a similar catchall, but is defined as being the correct type, which I do not think applies here.

I am not opposed to adding AttributeError to exceptions.py. Mostly the Idea is to have a base exception class from which all exceptions thrown by WrightTools itself can be caught. We did not define all of the standard lib exception, as we do not anticipate needing most of them.

Admittedly for the current scope of where interact2D is anticipated to be used, this kind of boxing of exceptions seems slightly superfluous. However, I'd like to have these systems in place and make sure it can be called in some larger application as well (perhaps as a pane in a larger GUI app)